pkg/cosign/tlog.go

```diff
@@ -93,10 +93,7 @@ func intotoEntry(ctx context.Context, signature, pubKey []byte) (models.Proposed
 // There are two Env variable that can be used to override this behaviour:
 // SIGSTORE_REKOR_PUBLIC_KEY - If specified, location of the file that contains
 // the Rekor Public Key on local filesystem
-// SIGSTORE_TRUST_REKOR_API_PUBLIC_KEY - If specified, fetches the Rekor public
-// key from the Rekor server using the provided rekorClient.
-// TODO: Rename SIGSTORE_TRUST_REKOR_API_PUBLIC_KEY to be test-only or remove.
-func GetRekorPubs(ctx context.Context, rekorClient *client.Rekor) (map[string]RekorPubKey, error) {
+func GetRekorPubs(ctx context.Context) (map[string]RekorPubKey, error) {
```

Comment:
Yes: I should document (and below) how users can trivially update

Reply:
Right, but doesn't that require a major version bump for a Go library? https://go.dev/doc/modules/release-workflow#breaking

Reply:
Yeah, I do know this but it's undocumented and unclear to me how I should do this.

https://github.com/sigstore/cosign/pull/2322 also included breaking API changes. Should I tag this with v1.14? It'd help if there was a suggestion here. Or are you hinting to me not to change the external API?

Reply:
> https://github.com/sigstore/cosign/pull/2322 also included breaking API changes.

Where? `cmd/` and `internal/` are exempt usually.

> Should I tag this with v1.14? It'd help if there was a suggestion here. Or are you hinting to me not to change the external API?

I don't know what I want 😭 But it'd have to be `v2`.

There's a two problems here running into each other:

1. Go as a language has unrealistic conventions for semver adherence, which *actually wind up mattering* in the tooling (Go will autoupdate you without thinking about it).
2. Cosign as a library didn't really give any thought to what should be public (see [IsTerminal](https://pkg.go.dev/github.com/sigstore/cosign@v1.13.1/pkg/cosign#IsTerminal)).

I think that these APIs are actually something people would use so we should err on the side of caution. 

Long-term, I think we should:

- Decouple the CLI version and the library version.
- Get in the habit of pushing new major library versions a lot more regularly. There's a perception I think of new major versions as something to be avoided, but this is exactly what they're for.
- Trim down the Cosign library API pretty aggressively. 

That's a lot to put on you for this PR. So for now maybe just leave them with unused parameters and a TODO? And I can file an issue about "strategy" and bring this up at a community meeting.

Reply:
> Where? cmd/ and internal/ are exempt usually.

In `pkg/`: Removal of `oci.DockerMediaTypesEnv`, and in `[pkg/providers/github/github.go](https://github.com/sigstore/cosign/pull/2322/files#diff-bc615284e09bac5bf71da5ed0ecf0d5e07104016844c9d64000e18accdcc9be7)`, removal of `github.RequestTokenEnvKey`,

Reply:
> That's a lot to put on you for this PR. So for now maybe just leave them with unused parameters and a TODO? And I can file an issue about "strategy" and bring this up at a community meeting.

Sure: this is overall a really hairy, annoying problem that is inevitably going to lead to inconsistencies. Do you know of any "detection" for semver miscompliance? It'd be nice to be bugged by a bot by this instead of relying on consistent reviews.

> Long-term, I think we should:

Completely agree! I'm waiting for `sigstore-go` on the front of trimming the library down too :)

Reply:
Ugh, you're totally right and I didn't see those even when I was looking specifically for them :(

Fortunately, I don't think anybody uses them, at least in OSS([DockerMediaTypesEnv](https://sourcegraph.com/search?q=context:global+DockerMediaTypesEnv&patternType=standard), [RequestTokenEnvKey](https://sourcegraph.com/search?q=context:global+RequestTokenEnvKey&patternType=standard), [RequestURLEnvKey](https://sourcegraph.com/search?q=context:global+RequestURLEnvKey&patternType=standard)). But yeah we should technically put them back.

I think that's not true for [GetRekorPubs](https://github.com/slsa-framework/slsa-verifier/blob/c151f4736825ab7b52187af27d8c844474bbc2ef/verifiers/internal/gha/rekor.go#L183)  and [VerifyTLogEntry](https://github.com/sigstore/scaffolding/blob/3420daff9cbf73f4e3fab17e21e8105c5260e788/cmd/prober/write.go#L153) (though between us we'd be able to fix the OSS ones pretty quick 😛).

> Do you know of any "detection" for semver miscompliance? It'd be nice to be bugged by a bot by this instead of relying on consistent reviews.

I couldn't agree with you more. I'll make sure to bring it up in any wider discussion.

Reply:
FYI https://github.com/sigstore/cosign/discussions/2365